### PR TITLE
Fix `OsuPlayfield` being retained indefinitely after gameplay

### PR DESCRIPTION
**Pull request by peppy**
Pointed out by @smoogipoo. Dropped everything to fix this one, as it was retaining one `OsuPlayfield` (~20mb) per play session.

There are multiple critical failures here, but the gist of it is that `ScoreInfo` holds `HitEvents` indefinitely, which holds `HitObjects` which has `DefaultsApplied` bound by `HitObjectLifetimeManager` which doesn't perform cleanup. One may argue that it should just perform cleanup and we can call it a day, but it's not straightforward (calling `Remove` on each `HitObject` will most definitely add an overhead) so I've instead TODO'd that and addressed the surrounding causes for reference retention.

The main retention path was the `ScorePerformanceCache` which has been addressed. The remaining safeties are just to avoid this happening in the 99% case.

Note that realm subscriptions still seem to hold references slightly longer than they should, but they eventually get finalised. I started to investigate this but the bar is high so that's for another day:

CleanShot 2024-01-29 at 06 48 30

CleanShot 2024-01-29 at 06 46 54

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Replays/OsuFramedReplayInputHandler.cs [hunk 1]
```diff
@@ -26,5 +26,12 @@ protected override void CollectReplayInputs(List<IInput> inputs)
             inputs.Add(new MousePositionAbsoluteInput { Position = GamefieldToScreenSpace(position) });
             inputs.Add(new ReplayState<OsuAction> { PressedActions = CurrentFrame?.Actions ?? new List<OsuAction>() });
         }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            GamefieldToScreenSpace = null;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Replays/OsuFramedReplayInputHandler.cs:34]: rather random change? is not explained in commit message or anything

Comment (peppy) [reply osu.Game.Rulesets.Osu/Replays/OsuFramedReplayInputHandler.cs:34]: Can probably be removed. Initially I was attempting to resolve the realm held paths, which this is one of.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Replays/OsuFramedReplayInputHandler.cs]: I've force pushed this out of the diff.

**Review by bdach — APPROVED** [review]

File: osu.Game/Screens/Ranking/ResultsScreen.cs [hunk 2]
```diff
@@ -275,6 +277,11 @@ public override bool OnExiting(ScreenExitEvent e)
             if (base.OnExiting(e))
                 return true;
 
+            // This is a stop-gap safety against components holding references to gameplay after exiting the gameplay flow.
+            // Right now, HitEvents are only used up to the results screen. If this changes in the future we need to remove
+            // HitObject references from HitEvent.
+            Score?.HitEvents.Clear();
```
Comment (smoogipoo) [inline osu.Game/Screens/Ranking/ResultsScreen.cs:283]: Given the rest of the PR, does this do anything meaningful at the moment, or is it just a potential safety?

Comment (peppy) [reply osu.Game/Screens/Ranking/ResultsScreen.cs:283]: See https://github.com/ppy/osu/pull/26770/commits/eb90ee5415b05dd827464fcd3c715f731da2df1f commit message ;)

**Pull request merged**